src/modules/socket.js

```diff
@@ -1,12 +1,186 @@
+  } else if (type === "payment" || type === "settlement") {
+    // 채팅 메시지 type이 "in"이거나 "out"인 경우 본문은 "${nickname} 님이 결제를 완료하였습니다" 또는 "${nickname} 님이 정산을 완료하였습니다"가 됩니다.
+    // TODO: 사용자 언어를 가져올 수 있다면 개선할 수 있다고 생각합니다.
+    const suffix =
+      type === "payment"
+        ? " 님이 결제를 완료하였습니다"
+        : " 님이 정산을 완료하였습니다";
+    return `${nickname} ${suffix}`;
+  }
+};
+
+/**
+ * 채팅을 전송하고 채팅 알림을 발생시킵니다.
+ * @summary express 라우터에서 채팅 이벤트를 보낼 수 있게 함수를 분리했습니다.
+ * @param {Server<DefaultEventsMap, DefaultEventsMap, DefaultEventsMap, any>} io - Socket.io 서버 인스턴스입니다. req.app.get("io")를 통해 접근할 수 있습니다.
+ * @param {Object} chat - 채팅 메시지 내용입니다.
+ * @param {string} chat.roomId - 채팅 및 채팅 알림을 보낼 방의 ObjectId입니다.
+ * @param {string} chat.type - 채팅 메시지의 유형입니다. "text" | "s3img" | "in" | "out" | "payment" | "settlement" 입니다.
+ * @param {string} chat.content - 채팅 메시지의 본문입니다. chat.type이 "s3img"인 경우에는 채팅의 objectId입니다. chat.type이 "in"이거나 "out"인 경우 입퇴장한 사용자의 id(!==ObjectId)입니다.
+ * @param {string} chat.authorId - 채팅을 보낸 사용자의 ObjectId입니다.
+ * @param {Date?} chat.time - optional. 채팅 메시지 전송 시각입니다.
+ * @return {Promise<Boolean>} 채팅 및 알림 전송에 성공하면 true, 중간에 오류가 발생하면 false를 반환합니다.
+ */
+const emitChatEvent = async (io, chat) => {
+  try {
+    // chat must contain type, content and authorId
+    // chat can contain time or not.
+    if (
+      !io ||
+      !chat?.roomId ||
+      !chat?.type ||
+      !chat?.content ||
+      !chat?.authorId
+    ) {
+      throw new IllegalArgumentsException();
+    }
+
+    const { roomId, type, content, authorId } = chat;
+    const time = chat?.time || Date.now();
+    const { nickname, profileImageUrl } = await userModel.findById(
+      authorId,
+      "nickname profileImageUrl"
+    );
+    if (!nickname) {
+      throw new IllegalArgumentsException();
+    }
+
+    const chatDocument = await chatModel
+      .findOneAndUpdate(
+        {
+          type,
+          authorId,
+          roomId,
+          time,
+        },
+        {
+          type,
+          authorId,
+          roomId,
+          time,
+          content,
+          isValid: true,
+        },
+        { upsert: true, new: true }
+      )
+      .lean();
+    chatDocument.authorName = nickname;
+    chatDocument.authorProfileUrl = profileImageUrl;
+
+    const room = await roomModel.findById(roomId, "name part");
+    const urlOnClick = `/myroom/${roomId}`;
+    const userIds = room.part.map((participant) => participant.user);
+    const userIdsExceptAuthor = room.part
+      .map((participant) => participant.user)
+      .filter((userId) => userId.toString() !== authorId.toString());
+    const deviceTokens = await getTokensOfUsers(userIdsExceptAuthor, {
+      chatting: true,
+    });
+
+    // 방의 모든 사용자에게 이미지 수신 이벤트를 발생시킵니다.
+    await Promise.all(
+      userIds.map(async (userId) =>
+        io.to(`user-${userId}`).emit("chat_push_back", {
+          chats: await transformChatsForRoom([chatDocument]),
+          roomId,
+        })
+      )
+    );
```

Comment:
```suggestion
    // 방의 모든 사용자에게 socket 메시지 수신 이벤트를 발생시킵니다.
    const chatsForRoom = await transformChatsForRoom([chatDocument]);
    await Promise.all(
      userIds.map(async (userId) =>
        io.to(`user-${userId}`).emit("chat_push_back", {
          chats: chatsForRoom,
          roomId,
        })
      )
    );
```

주석 설명이 잘못되었고, `transformChatsForRoom`를 `Promise.all` 밖으로 빼면 연산 수가 줄어들 것 같습니다 !